### PR TITLE
Enocde buildid+offset of unknown symbols in the frame

### PR DESCRIPTION
**Pull request by Jongy**

### CODE REVIEW

File: src/codeCache.h [hunk 1]
```diff
@@ -98,10 +100,11 @@ class NativeCodeCache : public CodeCache {
 
     void add(const void* start, int length, const char* name, bool update_bounds = false);
     void sort();
-    const char* binarySearch(const void* address);
+    const char* binarySearch(const void* address, LinearAllocator* allocator);
     const void* findSymbol(const char* name);
     const void* findSymbolByPrefix(const char* prefix);
     const void* findSymbolByPrefix(const char* prefix, int prefix_len);
+    void setBuildId(const char *build_id, int build_id_len);
```
Comment (YoniKF) [inline src/codeCache.h]:
```suggestion
    void setBuildId(const char* build_id, int build_id_len);
```
If you are interested in keeping conventions.

There are some other cases of this.

File: src/codeCache.cpp [hunk 2]
```diff
@@ -104,7 +107,19 @@ void NativeCodeCache::sort() {
...
-const char* NativeCodeCache::binarySearch(const void* address) {
+void NativeCodeCache::setBuildId(const char *build_id, int build_id_len) {
+    _build_id = (char*)malloc(build_id_len * 2 + 1);
+    if (_build_id) {
+        char *p = _build_id;
+        for (int i = 0; i < build_id_len; i++) {
+            p += sprintf(p, "%02hhx", build_id[i]);
+        }
+        // sprintf has null-terminated it for us
+        assert(p == _build_id + build_id_len * 2);
```
Comment (YoniKF) [inline src/codeCache.cpp]: If this code is loaded into the profiled process, better not to assert 👀.

**Review by YoniKF — COMMENTED** [review]
Cool!

File: src/codeCache.cpp (hunk 2 above)
Comment (Jongy) [reply src/codeCache.cpp]: Absolutely

File: src/codeCache.h (hunk 1 above)
Comment (Jongy) [reply src/codeCache.h]: Yeah, I tried to keep but seem to have forgotten it here :)

**Review by YoniKF — APPROVED** [review]

File: src/codeCache.cpp [hunk 3]
```diff
@@ -124,6 +138,19 @@ const char* NativeCodeCache::binarySearch(const void* address) {
...
     }
+
+    if (_add_build_ids && _build_id) {
+        char buf[512];
+        // add 1 for the null terminator
+        const int n = 1 + snprintf(buf, sizeof(buf), "%s %s+%p_[bid]", _name, _build_id, (void*)((char*)address - (char*)_min_address));
+        char *allocated;
+        // if snprintf didn't overflow, and we manage to allocate a suitable buffer (from the LinearAllocator, not malloc,
+        // since this is signal context and malloc can deadlock) - then use them.
+        if (n < sizeof(buf) && (allocated = (char*)allocator->alloc(n)) != NULL) {
```
Comment (Jongy) [inline src/codeCache.cpp:149]: I wanna do another round at verifying the lifetimes here (the pointer returned by `alloc()`)

File: src/arguments.h [hunk 4]
```diff
@@ -178,7 +179,8 @@ class Arguments {
         _end(NULL),
         _title("Flame Graph"),
         _minwidth(0),
-        _reverse(false) {
+        _reverse(false),
+        _build_ids(false) {
```
Comment (YoniKF) [inline src/arguments.h:183]: Didn't you want this to be enabled by default?
Or did you mean that gProfiler will pass this option by default?

File: src/codeCache.cpp (hunk 3 above)
Comment (YoniKF) [reply src/codeCache.cpp:149]: AFAICT, `LinearAllocator` is `clear`ed in two places:
* `LinearAllocator`'s destructor, which is called from `Profiler`'s destructor, which is a static variable so destructed only upon `async-profiler`'s unload .
* `CallTraceStorage::clear`, which is called from `Profiler::start`, which is called on a `start` command.

So I think you're good.

**Review by YoniKF — APPROVED** [review]

File: src/arguments.h (hunk 4 above)
Comment (YoniKF) [reply src/arguments.h:183]: Saw for myself in gProfiler code.

File: src/codeCache.cpp (hunk 3 above)
Comment (Jongy) [reply src/codeCache.cpp:149]: Sounds right. Thanks for the double check.

**Comment by Jongy** [comment]
I'll merge this once we're ready with the backend side.

**Pull request merged**